src/main/java/com/amazon/ion/impl/IonReaderContinuableCoreBinary.java

```diff
@@ -1369,10 +1378,301 @@ private enum State {
+            }
+            readValueAsExpression(expressions);
+        }
+
+        /**
+         * Reads a group expression.
+         * @param encoding how the group is encoded.
+         * @param expressions receives the expressions as they are materialized.
+         */
+        private void readGroupExpression(Macro.ParameterEncoding encoding, List<Expression.EExpressionBodyExpression> expressions) {
+            if (encoding == Macro.ParameterEncoding.Tagged) {
+                enterTaggedArgumentGroup();
+            } else {
+                enterTaglessArgumentGroup(encoding.taglessEncodingKind);
+            }
+            if (event == Event.NEEDS_DATA) {
+                throw new UnsupportedOperationException("TODO: support continuable parsing of macro arguments.");
+            }
+            int startIndex = expressions.size();
+            expressions.add(Expression.Placeholder.INSTANCE);
+            while (nextGroupedValue() != Event.NEEDS_INSTRUCTION) {
+                readValueAsExpression(expressions);
+            }
+            if (exitArgumentGroup() == Event.NEEDS_DATA) {
+                throw new UnsupportedOperationException("TODO: support continuable parsing of macro arguments.");
+            }
+            expressions.set(startIndex, new Expression.ExpressionGroup(startIndex, expressions.size()));
+        }
+
+        /**
+         * Reads a single parameter to a macro invocation.
+         * @param parameter information about the parameter from the macro signature.
+         * @param parameterPresence the presence bits dedicated to this parameter.
+         * @param expressions receives the expressions as they are materialized.
+         */
+        private void readParameter(Macro.Parameter parameter, long parameterPresence, List<Expression.EExpressionBodyExpression> expressions) {
+            Macro.ParameterEncoding encoding = parameter.getType();
+            switch (parameter.getCardinality()) {
+                case ZeroOrOne:
+                    if (parameterPresence == PresenceBitmap.EXPRESSION) {
+                        readSingleExpression(encoding, expressions);
+                    } else if (parameterPresence != PresenceBitmap.VOID) {
+                        throw new IllegalStateException();
+                    }
+                    break;
+                case ExactlyOne:
+                    readSingleExpression(encoding, expressions);
+                    break;
+                case OneOrMore:
+                    if (parameterPresence == PresenceBitmap.EXPRESSION) {
+                        readSingleExpression(encoding, expressions);
+                    }
+                    if (parameterPresence == PresenceBitmap.GROUP) {
+                        readGroupExpression(encoding, expressions);
+                    } else {
+                        throw new IllegalStateException();
```

Comment:
Minor suggestions to have a better error message.
```suggestion
                        throw new IonException("Invalid void argument for non-voidable parameter: " + parameter.getName());
```